flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java

```diff
@@ -259,28 +263,29 @@ public void restorePartitionedState(List<KeyGroupsStateHandle> state) throws Exc
 
 			try {
 
-				fsDataInputStream = keyGroupsHandle.getStateHandle().openInputStream();
+				fsDataInputStream = keyGroupsHandle.openInputStream();
 				cancelStreamRegistry.registerClosable(fsDataInputStream);
 
 				DataInputViewStreamWrapper inView = new DataInputViewStreamWrapper(fsDataInputStream);
 
 				int numKvStates = inView.readShort();
 
-				Map<Integer, String> kvStatesById = new HashMap<>(numKvStates);
-
 				for (int i = 0; i < numKvStates; ++i) {
 					String stateName = inView.readUTF();
 
-					TypeSerializer namespaceSerializer =
+					TypeSerializer<?> namespaceSerializer =
 							InstantiationUtil.deserializeObject(fsDataInputStream, userCodeClassLoader);
-					TypeSerializer stateSerializer =
+					TypeSerializer<?> stateSerializer =
 							InstantiationUtil.deserializeObject(fsDataInputStream, userCodeClassLoader);
 
-					StateTable<K, ?, ?> stateTable = new StateTable(stateSerializer,
-							namespaceSerializer,
-							keyGroupRange);
-					stateTables.put(stateName, stateTable);
-					kvStatesById.put(i, stateName);
+					StateTable<K, ?, ?> stateTable = stateTables.get(stateName);
+
+					if (null == stateTable) {
```

Comment:
Let's add a comment here that it is important to check for previously restored state first.